File: BOJ/24444-알고리즘 수업 - 너비 우선 탐색 1/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

var n, m, r int
var graph [][]int
var visited []int
var reader = bufio.NewReader(os.Stdin)
var writer = bufio.NewWriter(os.Stdout)

func input() {
	fmt.Fscan(reader, &n, &m, &r)
	graph = make([][]int, n+1)
	visited = make([]int, n+1)

	var u, v int
	for i := 0; i < m; i++ {
		fmt.Fscan(reader, &u, &v)
		graph[u] = append(graph[u], v)
		graph[v] = append(graph[v], u)
	}
	for i := 1; i <= n; i++ {
		sort.Ints(graph[i])
	}
}

func bfs(start int) {
	var queue []int
	count := 1
	queue = append(queue, start)

	for len(queue) != 0 {
		here := queue[0]
		queue = queue[1:]
		if visited[here] != 0 {
			continue
		}

		visited[here] = count
		count++

		for _, val := range graph[here] {
			if visited[val] == 0 {
				queue = append(queue, val)
			}
		}
	}
}

func solve() {
	bfs(r)
	for i := 1; i <= n; i++ {
		fmt.Fprintln(writer, visited[i])
	}
}

func main() {
	defer writer.Flush()
	input()
	solve()
}

File: BOJ/24444-알고리즘 수업 - 너비 우선 탐색 1/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int n, m, r;
vector<int> graph[100001];

void input() {
    cin >> n >> m >> r;
    
    int u, v;
    for (int i = 0; i < m; i++) {
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    for (int i = 1; i <= n; i++) sort(graph[i].begin(), graph[i].end());
}

void bfs() {
    int cnt = 1;
    queue<int> q;
    vector<int> visited(n + 1, 0);

    q.push(r); visited[r] = cnt++;
    while (!q.empty()) {
        int here = q.front(); q.pop();
        for (int i = 0; i < graph[here].size(); i++) {
            int next = graph[here][i];
            if (!visited[next]) {
                visited[next] = cnt++;
                q.push(next);    
            }
        }
    }
    for (int i = 1; i <= n; i++) cout << visited[i] << "\n";
}

int main(void) {
    cin.tie(0), ios::sync_with_stdio(0);
    input();
    bfs();
    return 0;
}